app/queries/servers/channel.ts

```diff
@@ -431,3 +432,102 @@ export function queryMyChannelsByUnread(database: Database, isUnread: boolean, s
+    let username = `u.username LIKE '${value}%'`;
+    let displayname = `c.display_name LIKE '${value}%'`;
+    if (!matchStart) {
+        username = `u.username LIKE '%${value}%' AND u.username NOT LIKE '${value}%'`;
+        displayname = `(c.display_name LIKE '%${value}%' AND c.display_name NOT LIKE '${value}%')`;
+    }
+    const currentUserId = observeCurrentUserId(database);
+    return currentUserId.pipe(
+        switchMap((uId) => {
+            return database.get<MyChannelModel>(CHANNEL).query(
+                Q.unsafeSqlQuery(`SELECT DISTINCT my.* FROM ${MY_CHANNEL} my
+                INNER JOIN ${CHANNEL} c ON c.id=my.id AND c.team_id='' AND c.delete_at=0 ${onlyDMs}
+                INNER JOIN ${CHANNEL_MEMBERSHIP} cm ON cm.channel_id=my.id
+                INNER JOIN ${USER} u ON u.id=cm.user_id AND (cm.user_id != '${uId}' AND ${username})
+                OR ${displayname}
+                ORDER BY my.last_viewed_at DESC
+                LIMIT ${take}`),
+            ).observe();
+        }),
+    );
+};
+
+export const observeNotDirectChannelsByTerm = (database: Database, term: string, take = 20, matchStart = false) => {
+    const teammateNameSetting = observeTeammateNameDisplay(database);
+
+    const value = Q.sanitizeLikeString(term.startsWith('@') ? term.substring(1) : term);
+    let username = `u.username LIKE '${value}%'`;
+    let nickname = `u.nickname LIKE '${value}%'`;
+    let displayname = `(u.first_name || ' ' || u.last_name) LIKE '${value}%'`;
+    if (!matchStart) {
+        username = `(u.username LIKE '%${value}%' AND u.username NOT LIKE '${value}%')`;
+        nickname = `(u.nickname LIKE '%${value}%' AND u.nickname NOT LIKE '${value}%')`;
+        displayname = `((u.first_name || ' ' || u.last_name) LIKE '%${value}%' AND (u.first_name || ' ' || u.last_name) NOT LIKE '${value}%')`;
+    }
+
+    return teammateNameSetting.pipe(
+        switchMap((setting) => {
+            let sortBy = '';
+            switch (setting) {
+                case General.TEAMMATE_NAME_DISPLAY.SHOW_NICKNAME_FULLNAME:
+                    sortBy = "ORDER BY CASE u.nickname WHEN '' THEN 1 ELSE 0 END, CASE (u.first_name || ' ' || u.last_name) WHEN '' THEN 1 ELSE 0 END, u.username";
+                    break;
+                case General.TEAMMATE_NAME_DISPLAY.SHOW_FULLNAME:
+                    sortBy = "ORDER BY CASE (u.first_name || ' ' || u.last_name) WHEN '' THEN 1 ELSE 0 END, u.username";
+                    break;
+                default:
+                    sortBy = 'ORDER BY u.username';
+                    break;
+            }
+
+            return database.get<UserModel>(USER).query(
+                Q.unsafeSqlQuery(`SELECT DISTINCT u.* FROM User u
+                LEFT JOIN ChannelMembership cm ON cm.user_id=u.id
+                WHERE cm.user_id IS NULL AND (${displayname} OR ${username} OR ${nickname})
+                ${sortBy} LIMIT ${take}`),
+            ).observe();
```

Comment:
Second round understanding what this is doing :sweat_smile:

So... this gets all users, join with the ChannelMembership table, and get all the users that don't have a membership.
But wouldn't that mean that you are taking only the users that are not in any shared channel with you? Not only the ones that don't have a direct channel with you?

Reply:
that is a great catch! I'm going to check that this matches DM channels.